File: src/main/java/xyz/mlhmz/financemanager/entities/Transaction.java
package xyz.mlhmz.financemanager.entities;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;
import org.hibernate.annotations.UuidGenerator;

import java.time.LocalDateTime;
import java.util.UUID;

@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
@Builder
@Entity
public class Transaction {
    @Id
    @UuidGenerator
    private UUID uuid;

    private String title;

    private String description;

    private Double amount;

    private LocalDateTime timestamp;

    @CreationTimestamp
    private LocalDateTime createdAt;

    @UpdateTimestamp
    private LocalDateTime updatedAt;

    @ManyToOne
    @JoinTable(name = "category_transactions", joinColumns = {@JoinColumn(name = "transactions_uuid", referencedColumnName = "uuid")},
            inverseJoinColumns={@JoinColumn(name="category_uuid", referencedColumnName="uuid")})
    private Category category;

    @ManyToOne
    @JoinTable(name = "sheet_transactions", joinColumns = {@JoinColumn(name = "transactions_uuid", referencedColumnName = "uuid")},
            inverseJoinColumns={@JoinColumn(name="sheet_uuid", referencedColumnName="uuid")})
    private Sheet sheet;

    @ManyToOne(optional = false)
    private OAuthUser user;
}

File: frontend/src/transactions/Transaction.ts
import { z } from "zod";
import { Category } from "../categories/Category";

export interface Transaction {
	uuid?: string;
	title?: string;
	description?: string;
	amount?: number;
	timestamp?: string;
	createdAt?: string;
	updatedAt?: string;
	category?: Category;
	sheet?: TransactionSheet;
}

export interface TransactionSheet {
	uuid?: string;
	title?: string;
}

export const MutateTransaction = z.object({
	title: z.string(),
	description: z.string().optional(),
	amount: z.number(),
	timestamp: z.date(),
	sheetId: z.string().uuid(),
	categoryId: z.string().uuid(),
});

export type MutateTransaction = z.infer<typeof MutateTransaction>;
